bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java

```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
                 targetTimedEffects.setDuration(duration);
             }
         }
-
         return target;
     }
+
+    /**
+     * Merge on/dimming/color fields from light and grouped light resources.
+     *
+     * Subsequent sparse resources will be merged into the first one.
+     * Sparse resources will be merged into any prior full state resources.
+     * Full state resources will replace any prior sparse resources.
+     */
+    public static Collection<Resource> mergeLightResources(Collection<Resource> resources) {
+        Map<String, Resource> resourceIndex = new HashMap<>();
+        Iterator<Resource> iterator = resources.iterator();
+        while (iterator.hasNext()) {
+            Resource resource = iterator.next();
+            String id = resource.getId();
+            Resource indexedResource = resourceIndex.get(id);
+            if (indexedResource == null) {
+                indexedResource = resource;
+                resourceIndex.put(id, indexedResource);
+            }
+
+            if (!LIGHT_TYPES.contains(resource.getType()) || !resource.hasHSBComponent()
+                    || resource.equals(indexedResource)) {
+                continue;
+            }
+
+            // Full state takes precedence, invalidate previous sparse resources.
+            if (resource.hasFullState()) {
+                indexedResource.setOnState(null);
+                indexedResource.setDimming(null);
+                indexedResource.setColorXy(null);
+
+                resourceIndex.put(id, resource);
+                continue;
+            }
+
+            OnState onState = resource.getOnState();
+            if (onState != null) {
+                indexedResource.setOnState(onState);
+                resource.setOnState(null);
+            }
+            Dimming dimming = resource.getDimming();
+            if (dimming != null) {
+                indexedResource.setDimming(dimming);
+                resource.setDimming(null);
+            }
+            ColorXy colorXy = resource.getColorXy();
+            if (colorXy != null) {
+                indexedResource.setColorXy(colorXy);
+                resource.setColorXy(null);
+            }
+
+            if (!resource.hasItems()) {
```

Comment:
I am not sure if I agree with your approach here.

Your approach (finally) is to de-link the 'hsbFields' references from resources later in the list and re-link them to the first resource in the list. (Which is actually what I was proposing). This is good. But this means that any references to 'nonHsbFields' are left behind ('orphaned') in the later resource. This is where I have a doubt.

My feeling is that if the resources have ONLY 'hsbFields' and NO 'nonHsbFields' then one can indeed safely de-/re-link the 'hsbFields' references, and remove the now empty resource from the list. HOWEVER if the resources have BOTH 'hsbFields' and ALSO 'nonHsbFields' (a 'complex' resource), then it is better NOT to de-/re-link any of the field references; and rather let both resources pass through verbatim in this list. My logic is that such a 'complex' resource with mixed fields may contain perhaps more complex logic, and I think we should rather not mess around with such complex objects.

Reply:
The only reason for merging anything rather than keeping it separate and processing it sequentially (like before this PR) is because we have a dependency between **on**, **dimming** and **color** on the openHAB side because of the combined HSB.

Therefore the objective is to mitigate multiple state updates by merging such resources. For anything else, there is strictly no need to merge, so therefore I kept the logic strictly to fulfil this requirement.

I also checked `updateChannels`, and there isn't any problems with this, it simply processes the objects that exists. This also makes perfect sense, because we receive sparse resources and should be able to handle them individually already.

Reply:
> we have a dependency between on, dimming and color .. Therefore the objective is to mitigate multiple state updates by merging such resources.
> For anything else, there is strictly no need to merge, so therefore I kept the logic strictly to fulfil this requirement.

Indeed. My point relates to your "anything else": If either of the resources has on, dimming and color and ALSO **_anything else_** then I think there may be a risk (now or future) if we mess around with the resources. So my suggestion is as follows..
- Both resources have on, dimming and color ONLY => merge Ok
- Either resource has on, dimming and color and ALSO **_anything else_** => do not merge

Reply:
> To be more precise, I fear that the 'merge some fields but not others' (half pregnant) approach may be risky..

Why do you think that? There should be no dependencies between the objects, which we already process individually as we receive them. The only dependency is the one we have because of HSB. There shouldn't be any difference between us separating them apart and receiving them individually straight from the source, i.e. it would be quite unexpected that this would cause any problems. Can you please give an example?

> Either resource has on, dimming and color and ALSO anything else => do not merge

That would break the fix. Simplest example:
- Resource A: on=on
- Resource B: dimming=75, effects=something

I don't know if this can happen, but I wouldn't make any assumptions that it can't, and also that's not the point. With current logic we'll process:
- Resource A: on=on, dimming=75
- Resource B: effects=something

This will work. If we wouldn't merge in this case, we would still have the bug under those circumstances.

Reply:
> I don't know if this can happen, but I wouldn't make any assumptions that it can't

Me neither. That IS my point ..

> This will work. 

Are you not being over-confident? The [API](https://developers.meethue.com/develop/hue-api-v2/api-reference/#resource_light_get) for LIGHT includes many elements, and I am simply wondering if either now or in future the bridge may pack combined 'complex' elements into a single specific resource so that they may be read in combination (I am thinking of say dynamics, alert, mode, signalling or gradient (some of which we don't currently use) which might need to be interpreted in combination with a colorxy, dimming, or onoff element; and if we had 'stolen' that element from resource #2 to put in resource #1, then the resource #2 resource might lose its meaning.

I am just worried that by doing something non transparent here and now, we might be creating potentially a quite hard to debug problem later.

However to take the other side of the argument ;) -- I have always argued that Hue DTOs are orthogonal, so the other side of the argument would be to not expect any such non orthogonality and not be upset if any non orthogonality would be broken by this fix.